### PR TITLE
Share notebook's metadata

### PR DESCRIPTION
**Pull request by hbcarlos**
As discussed in #11038, by default the metadata is not shared because there are some attributes that we do not want to share. This makes every use of the metadata in the JupyterLab, not collaborative. In the case you want to share the metadata, you need to explicitly add the content to the shared model instead of modelDB. With this PR, we listen to changes in modelDB and apply them to the shared model to make everything collaborative.

In this PR we are facing one small problem. ModelDB allows modifying the attributes independently while the shared model doesn't. To update the modelDB on a remote client we need to update the whole metadata because is the update we receive from the shared model. Maybe we should change how we update the metadata in the shared model.

## References
Solves #11038

## Code changes
Listening for changes on modelDB metadata to update the shared model. The problem with this approach is that we need to update the modelDB on other clients and that triggers the change event. For this reason, we create some sort of a mutex to avoid updating again the client.

## User-facing changes
Now the metadata is shared between clients.

### CODE REVIEW

File: packages/notebook/src/model.ts [hunk 1]
```diff
@@ -428,6 +429,27 @@ close the notebook without saving it.`,
...
+        });
+      });
+    }
+  }
+
+  private _onMetadataChanged(
+    metadata: IObservableJSON,
+    change: IObservableMap.IChangedArgs<ReadonlyPartialJSONValue | undefined>
+  ): void {
+    if (change.key !== 'kernelspec' && change.key !== 'language_info') {
```
Comment (fcollonval) [inline packages/notebook/src/model.ts:447]: I would set that as constant at the module level:

```suggestion
    if ( !UNSHARED_KEYS.includes(change.key) ) {
```

**Review by fcollonval — COMMENTED** [review]
I have only one suggestion to set the _unshared_ keys in a constant tuple at the module level.

File: packages/notebook/src/model.ts (hunk 1 above)
Comment (hbcarlos) [reply packages/notebook/src/model.ts]: Hey @fcollonval, thanks for the suggestion. I applied the changes on my last commit.

**Review by fcollonval — APPROVED** [review]
LGTM

**Comment by blink1073** [comment]
@meeseeksdev please backport to 3.2.x

**Comment by blink1073** [comment]
@meeseeksdev please backport to 3.1.x

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/jupyterhub-collaboration-and-tags/11259/3

**Pull request merged**